### ISSUE TITLE
Overall comments on the content of the repository

### ISSUE THREAD

**Issue by floridop** [felixffjf/MNXB11-Project-TeamB#10]
This issue contains feedback about your project. In some cases it may 
contain actions to carry in order to get a final grade.

# Workplan

The document is minimalistic, and contains no deadlines or checkpoints,
overall seems it was used only to agree on tasks.

# ChangeLog

There is no ChangeLog.

# README.md

The document is very detailed but it doesn't really clarify what the
dependencies are. There are several typos.

It boldly suggests that everything is as simple as
running `make` but:

```shell
[pflorido@cosmos-pkf TeamB]$ make
tar zxvf datasets/datasets.tgz -C datasets
smhi-opendata_1_105370_20231007_154742_Falun.csv
smhi-opendata_1_140480_20231007_155326_Umea.csv
smhi-opendata_1_162860_20231007_155220_Lulea.csv
smhi-opendata_1_52230_20231007_155448_Falsterbo.csv
smhi-opendata_1_53430_20231007_155558_Lund.csv
smhi-opendata_1_62410_20231007_160006_Halmstad.csv
smhi-opendata_1_72450_20231007_155620_Boras.csv
smhi-opendata_1_78400_20231007_155644_Visby.csv
smhi-opendata_1_93220_20231007_155708_Karlstad.csv
smhi-opendata_1_97530_20231007_155803_Uppsala.csv
smhi-opendata_1_99450_20231007_155728_Soderarm.csv
uppsala_tm_1722-2022.dat
uppsala_tm_1722-2022.txt
bash src/CleanerAndFilter/filter1950-2023.sh datasets/smhi-opendata_1_162860_20231007_155220_Lulea.csv
Redirecting filter logs to 2024-11-16_filter1950-2023.sh.log
[2024-11-16T13:06:26+01:00 Filter]: Calling shmicleaner.sh script
Redirecting cleaner logs to 2024-11-16_smhicleaner.sh.log
[2024-11-16T13:06:26+01:00 Cleaner]: Copying input file datasets/smhi-opendata_1_162860_20231007_155220_Lulea.csv to original_smhi-opendata_1_162860_20231007_155220_Lulea.csv
[2024-11-16T13:06:26+01:00 Cleaner]: Finding the first line containing 'Datum'...
[2024-11-16T13:06:26+01:00 Cleaner]: Found line 10
[2024-11-16T13:06:26+01:00 Cleaner]: Perform cleanup in one line, result in baredata_smhi-opendata_1_162860_20231007_155220_Lulea.csv
[2024-11-16T13:06:26+01:00 Filter]: Only saving measurements taken between 1950 and 2023, writing to Lulea
Filtered output saved to Lulea
g++ -Wall -Wextra -Werror -std=c++17 `root-config --cflags --libs` -o src/TempDifference/yearly_temp_difference src/TempDifference/yearly_temp_difference.cxx
/bin/sh: line 1: root-config: command not found
src/TempDifference/yearly_temp_difference.cxx:9:10: fatal error: TCanvas.h: No such file or directory
    9 | #include "TCanvas.h"
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:64: src/TempDifference/yearly_temp_difference] Error 1
```

A better README should have given instructions how to install ROOT or
start the course container.

This is only suggested, but never described much later in the document at https://github.com/felixffjf/MNXB11-Project-TeamB?tab=readme-ov-file#step-1-set-up-your-environment.

Anyway I got it and I ran everything inside the container.

the make step and compilation all went smooth. Good job!

I didn't really understand the sentence at line: 
https://github.com/felixffjf/MNXB11-Project-TeamB/blob/d801b87134bc70e886670892e447fc62a6f04f28/README.md?plain=1#L31

> This project has been made in the course Apptainer which already had ROOT and all the necceary tools except the libraries mentioned in the begining of the file. 

Which file? The `Makefile`? I don't know. Which libraries? I am guessing you meant the libraries referred by the `include` statements in each C++ related file.

The rest of the README tries to explain how to create a project, I don't really understand how this shows what you did and how what you did works, but nice attempt.

I also disagree that one needs to use exaclty the includes you suggest. Makefiles and make are usually agnostic of what one uses as long as a proper include and
library folder is provided, which is not the case with your Makefile since you confused it with a scripting tool.

It makes more sense to use the Makefile to manage include and library folders and compilation targets rather than a scripting tool.

Overall I do not really like the "How to add your own project" section  because you mix editing the Makefile with lines of code that should be in .cxx and it gets very confusing.

I appreciate the attempt.

# BASH Code

I very much appreciated that you placed the bash code inside the `src/` folder.
 
`smhicleaner.sh` was pretty much the same provided for the case study.

`filter` is definitely inspired by the provided `smhifilter.sh` in the case study with your customization, which I tought is was nice.

The overall code looks ok, and the filter as we discusses was probably inspired by ChatGPT due to its complexity.

However there are a few bad choices that I want to highlight:

In Line 58 you hardcode the path to `smhicleaner.sh`
https://github.com/felixffjf/MNXB11-Project-TeamB/blob/d801b87134bc70e886670892e447fc62a6f04f28/src/CleanerAndFilter/filter1950-2023.sh#L58

This is quite bad because it implies that one runs `filter1950-2023.sh` from the root folder of your project, which in turn means
that your makefile will only work if called in the root of your project, but this is not necessarily the case. `make` can be
executed from everywhere and the Makefile must make sure the relative paths are resolved. 

In practice your scripts only work when called from the root of the project and with the exact folder structure 
of your project. I do not think this was necessary and it implies you need to change both the Makefile AND the script if
you move the scripts somewhere else (for example a `scripts/` folder). It is not portable.

As I said I did not like that you remove the extension from csv files:
https://github.com/felixffjf/MNXB11-Project-TeamB/blob/d801b87134bc70e886670892e447fc62a6f04f28/src/CleanerAndFilter/filter1950-2023.sh#L89

Be nice and let the user know what kind of output data file you're producing.

# C++ Code

Looking at the code I confirm a feeling we had during the presentation,
that is, it's quite sad that the outcome of teamwork became copypaste.

Templating is ok, I understand you used each other's code, but this
is really not scalable. 3 different main functions doesn't really make
groupwork I am afraid. You code is not reusable if not by copy pasting
everything. In other words it kind of matches the workplan where there 
is no description of a possible architecture for the code.

The other sad thing is that you didn't use any object oriented feature
but ROOT classes, which is also strange given that your group work
would have benefit from at least one common class for the data,
with all the functionalities requested.

You seem to have an adequate competence in C++ syntax and coding style,
but you did not prove you can design object oriented programming code.

I leave further comments to @joachimckh .

# ROOT

I appreciated that you used root as a C++ library instead of using ROOT scripts.

# Summary

It is very nice that you automated all the compilation and execution of
the project. However you used the wrong tool, Makefiles are to automate
compilation and binary generation and not processing.

To automate processing you should use a BASH script and parametrize
the input, instead in your case you're forced to add a new variable
to the makefile and edit it for every data file, which is crazy if
you had to process thousands of different data files from different cities.

In other words, separate the static tasks that can be done by make (compiling a binary)
from dynamic tasks that depend on the input (cleanup dataset, launch binaries on certain input)
and pass dynamic parameters via command line istead of editing a huge Makefile.

I do not like that you save out files all over the place and that you even save
the cleaned .csv files without extension (see the `Lund Lulea` cleaned files)
Be nice with the user and your fellow colleagues. Do not remove information
when it's not necessary. Your output were .csv datafiles and it would have
been good to (1) keep the file extension and (2) describe a bit what the file contained in the name, such as `cleaned_Lund.csv`

# Action(s) to take

No action to take.
